common/step_download.go

```diff
@@ -90,19 +92,34 @@ func (s *StepDownload) Run(state multistep.StateBag) multistep.StepAction {
 			Checksum:   checksum,
 			UserAgent:  "Packer",
 		}
+		downloadConfigs[i] = config
 
-		path, err, retry := s.download(config, state)
-		if err != nil {
-			ui.Message(fmt.Sprintf("Error downloading: %s", err))
+		if match, _ := NewDownloadClient(config).VerifyChecksum(config.TargetPath); match {
+			ui.Message(fmt.Sprintf("Found already downloaded, initial checksum matched, no download needed: %s", url))
+			finalPath = config.TargetPath
+			break
 		}
+	}
 
-		if !retry {
-			return multistep.ActionHalt
-		}
+	if finalPath == "" {
```

Comment:
Why add a second loop here and introduce the need to track the download configs?  Why not just do the check and then the download inside the same loop?  This feels like unnecessary complexity on a first read.

Reply:
The whole PR is about adding this second loop. Maybe I've poorly described problem. 
So imagine next case:
1. There're two urls in 'iso_urls', first one accessible only in internal network (link to local cache actually), second is accessible from anywhere.
2. Running packer build from place where first link doesn't work, got file downloaded from second one and saved as `packer_cache/some_hash_of_second_url.iso`
3. Then running packer in place where first link is accessible, got packer re-downloaded file and stored it as `packer_cache/some_hash_of_first_url.iso`. But obviously packer could have use `packer_cache/some_hash_of_second_url.iso`.

So my PR is to change behaviour on third step, to check for ALL downloaded files prior to ANY download.

Reply:
Ah you're right I read this too fast and misunderstood the goal.